src/activity_handlers.cpp

```diff
@@ -2637,28 +2639,27 @@ void activity_handlers::craft_do_turn( player_activity *act, player *p )
         p->cancel_activity();
         return;
     }
-    act->moves_left -= crafting_speed * p->get_moves();
-    p->set_moves( 0 );
     if( calendar::once_every( 1_hours ) && crafting_speed < 0.75f ) {
         // TODO: Describe the causes of slowdown
         p->add_msg_if_player( m_bad, _( "You can't focus and are working slowly." ) );
     }
-}
 
-void activity_handlers::craft_finish( player_activity *act, player *p )
-{
-    p->complete_craft();
+    craft.item_counter += crafting_speed * p->get_moves();
+    p->set_moves( 0 );
     act->set_to_null();
-}
 
-void activity_handlers::longcraft_finish( player_activity *act, player *p )
-{
-    const int batch_size = act->values.front();
-    p->complete_craft();
-    act->set_to_null();
-    // Workaround for a bug where longcraft can be unset in complete_craft().
-    if( p->making_would_work( p->lastrecipe, batch_size ) ) {
-        p->last_craft->execute();
+    if( craft.item_counter >= rec.time ) {
+        item craft_copy = craft;
+        p->i_rem( pos );
+        p->complete_craft( craft_copy );
+        if( act->values[1] ) { // If this is a long craft
+            if( p->making_would_work( p->lastrecipe, craft_copy.charges ) ) {
+                p->last_craft->execute();
+            }
+        }
+    } else {
+        p->assign_activity( activity_id( "ACT_CRAFT" ) );
```

Comment:
Why restarting the activity?

Reply:
This is standard practice with activities where it isn't known how long they will take.  See `ACT_MOVE_ITEMS`, `ACT_DROP`, or `ACT_MOVE_LOOT` for example.

Reply:
You're not force to do so. Moreover, re-starting the activity each turn looks hacky. The alternative (and cleaner) approach is invoking `cancel_activity` when you don't want to proceed (`ACT_DROP` does exactly this).

Reply:
Oh you're totally right, thanks!